File: tic_tac_poe/src/board.py
class Board:
    def __init__(self, size=3):
        self.size = size
        self.cells = [["   "] * size for _ in range(size)]

    def get_size(self):
        return self.size

    def get_cells(self):
        return self.cells

    def is_valid_position(self, row, col):
        if row < 0 or row >= self.size or col < 0 or col >= self.size:
            return False

        if self.cells[row][col].strip() != "":
            return False

        return True

    def set_cell(self, row, col, mark):
        if not self.is_valid_position(row, col):
            raise ValueError("Invalid row or column value. Try again")

        self.cells[row][col] = f" {mark.value} "

    def display(self):
        print("      c1  c2  c3")
        print("    -------------")

        for i in range(len(self.cells)):
            print("r{}  |".format(i + 1), end="")

            for j in range(len(self.cells[i])):
                print(self.cells[i][j], end="|")

            print()

            if i < 2:
                print("    -------------")
            print("    -------------")

    def is_full(self):
        return all(cell.strip() not in ("", "-") for row in self.cells for cell in row)

    def has_horizontal_win(self):
        for row in self.cells:
            trimmed_row = [cell.strip() for cell in row]
            if len(set(trimmed_row)) == 1 and "" not in trimmed_row:
                return True
        return False

    def has_vertical_win(self):
        for col in zip(*self.cells):
            trimmed_col = [cell.strip() for cell in col]
            if len(set(trimmed_col)) == 1 and "" not in trimmed_col:
                return True
        return False

    def has_diagonal_win(self):
        if self.cells[0][0].strip() != "" and all(self.cells[i][i].strip() == self.cells[0][0].strip() for i in range(self.size)):
            return True
        if self.cells[0][self.size - 1].strip() != "" and all(self.cells[i][self.size - i - 1].strip() == self.cells[0][self.size - 1].strip() for i in range(self.size)):
            return True
        return False

    def check_winner(self):
        return self.has_horizontal_win() or self.has_vertical_win() or self.has_diagonal_win()

    def set_remaining_cells(self):
        self.cells[:] = [[" - " if cell.strip() == "" else cell for cell in row]
                         for row in self.cells]

File: tic_tac_rs/src/board.rs
use crate::utils::Mark;

pub trait BoardTrait {
    fn get_size(&self) -> &usize;
    fn get_cells(&self) -> &Vec<Vec<String>>;
    fn is_valid_position(&self, row: usize, col: usize) -> bool;
    fn get_cell(&self, row: usize, col: usize) -> &str;
    fn set_cell(&mut self, row: usize, col: usize, mark: Mark);
    fn display(&self);
    fn is_full(&self) -> bool;
    fn has_horizontal_win(&self) -> bool;
    fn has_vertical_win(&self) -> bool;
    fn has_diagonal_win(&self) -> bool;
    fn check_winner(&self) -> bool;
    fn set_remaining_cells(&mut self);
}

#[derive(Clone)]
pub struct Board {
    pub cells: Vec<Vec<String>>,
    pub size: usize
}

impl Board {
    pub fn new(size: Option<usize>) -> Self {
        let size = size.unwrap_or(3);
        let cells = vec![vec!["   ".to_string(); size]; size];
        Board { cells, size }
    }
}

impl BoardTrait for Board {
    fn get_size(&self) -> &usize {
        &self.size
    }

    fn get_cells(&self) -> &Vec<Vec<String>> {
        &self.cells
    }

    fn is_valid_position(&self, row: usize, col: usize) -> bool {
        if row >= self.size || col >= self.size {
            return false;
        }
        if !self.cells[row][col].trim().is_empty() {
            return false;
        }
        return true;
    }

    fn get_cell(&self, row: usize, col: usize) -> &str {
        if !self.is_valid_position(row, col) {
            panic!("Invalid row or column value. Try again");
        }
        &self.cells[row][col].as_str()
    }

    fn set_cell(&mut self, row: usize, col: usize, mark: Mark) {
        if !self.is_valid_position(row, col) {
            panic!("Invalid row or column value. Try again");
        }

        self.cells[row][col] = format!(" {} ", mark);
    }

    fn display(&self) {
        println!("      c1  c2  c3");
        println!("    -------------");

        for i in 0..self.cells.len() {
            print!("r{}  |", i + 1);

            for j in 0..self.cells[i].len() {
                print!("{}|", self.cells[i][j]);
            }

            println!();

            if i < 2 {
                println!("    -------------");
            }
        }

        println!("    -------------");
    }

    fn is_full(&self) -> bool {
        self.cells.iter().all(|row| {
            row.iter().all(|cell| {
                let trimmed_cell = cell.trim();
                trimmed_cell != "" && trimmed_cell != "-"
            })
        })
    }

    fn has_horizontal_win(&self) -> bool {
        for row in &self.cells {
            let trimmed_row: Vec<&str> = row.iter().map(|cell| cell.trim()).collect();

            if trimmed_row.iter().all(|&cell| cell != "" && cell == trimmed_row[0])  {
                return true;
            }
        }

        false
    }

    fn has_vertical_win(&self) -> bool {
        for col in (0..self.size).map(|i| self.cells.iter().map(|row| &row[i]).collect::<Vec<_>>()) {
            let trimmed_col: Vec<&str> = col.iter().map(|cell| cell.trim()).collect();

            if trimmed_col.iter().all(|&cell| cell != "" && cell == trimmed_col[0])  {
                return true;
            }
        }

        false
    }

    fn has_diagonal_win(&self) -> bool {
        if self.cells[0][0].trim() != "" && (0..self.size).all(|i| self.cells[i][i].trim() == self.cells[0][0].trim()) {
            return true
        }

        if self.cells[0][self.size - 1].trim() != "" && (0..self.size).all(|i| self.cells[i][self.size - i - 1].trim() == self.cells[0][self.size - 1].trim()) {
            return true
        }

        false
    }

    fn check_winner(&self) -> bool {
        self.has_horizontal_win() || self.has_vertical_win() || self.has_diagonal_win()
    }

    fn set_remaining_cells(&mut self) {
        for row in self.cells.iter_mut() {
            *row = row.iter().map(|cell| {
                if cell.trim() == "" {
                    " - ".to_string()
                } else {
                    cell.to_string()
                }
            }).collect::<Vec<_>>();
        }
    }

}

